### DISCUSSION TITLE
Changing Twig compiler

### DISCUSSION THREAD

**Discussion by seferov** [symfony/symfony#43033] [Q&A] [answered]
I want to change [twig compiler](https://github.com/twigphp/Twig/blob/464842c41f63254eaf06a26489b773b956127f9d/src/Environment.php#L482). Is there any way do it within Symfony framework? Twig service description:

```xml
<service id="twig" class="Twig\Environment" public="true">
    <argument type="service" id="twig.loader" />
    <argument /> <!-- Twig options -->
    <call method="addGlobal">
        <argument>app</argument>
        <argument type="service" id="twig.app_variable" />
    </call>
    <call method="addRuntimeLoader">
        <argument type="service" id="twig.runtime_loader" />
    </call>
    <configurator service="twig.configurator.environment" method="configure" />
    <tag name="container.preload" class="Twig\Cache\FilesystemCache" />
    <tag name="container.preload" class="Twig\Extension\CoreExtension" />
    <tag name="container.preload" class="Twig\Extension\EscaperExtension" />
    <tag name="container.preload" class="Twig\Extension\OptimizerExtension" />
    <tag name="container.preload" class="Twig\Extension\StagingExtension" />
    <tag name="container.preload" class="Twig\ExtensionSet" />
    <tag name="container.preload" class="Twig\Template" />
    <tag name="container.preload" class="Twig\TemplateWrapper" />
</service>
```

Making it through setter method seems cannot work because Compiler class needs twig itself for construction thus results in circular reference. My goal is to change twig compiler with a custom one before Symfony cache warmup.

**Comment by stof** [comment]
what is your use case for changing the Compiler ?

**Reply by seferov** [reply]
@stof I want to change behavior of `twig_get_attribute` function. When strict check is disabled, instead of early return I want to catch RuntimeError (don't throw exception) and log the error messages.

**Reply by stof** [reply]
I don't even think changing the compiler will allow you to do that. The compiler is not the one implementing the `twig_get_attribute` function.

**Reply by seferov** [reply]
yes, it might be hacky but in the compiler I can replace `twig_get_attribute` with my custom function. I cannot find another way to achieve what I want.

**Comment by javiereguiluz** [comment, marked as answer]
Maybe you can do something with a custom base `Template` class? I've never seen this option used anywhere, but maybe it's helpful for your use case: https://symfony.com/doc/current/reference/configuration/twig.html#base-template-class

**Reply by seferov** [reply]
thank you for the answer. I think this approach might solve my problem
